ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Background.hs

```diff
@@ -531,7 +532,14 @@ addBlockRunner
 addBlockRunner cdb@CDB{..} = forever $ do
     let trace = traceWith cdbTracer . TraceAddBlockEvent
     trace $ PoppedBlockFromQueue RisingEdge
-    blkToAdd <- getBlockToAdd cdbBlocksToAdd
-    trace $ PoppedBlockFromQueue $ FallingEdgeWith $
-      blockRealPoint $ blockToAdd blkToAdd
-    addBlockSync cdb blkToAdd
+    -- if the `addBlockSync` does not complete because it was killed by an async
+    -- exception (or it errored), notify the blocked thread
```

Comment:
Here we notify the thread that was waiting for this block, but why do we need to notify all other blocks and flush the queue? later blocks could be tried later on, no need to drop them

Reply:
If the thread is killed by and async exception there won't be a chance for them to be tried later on. So it is a good thing to flush the queue and notify all others that are blocked so we don't lose those.

There could be a potential race and some other thread add to the queue after we flush but it will be rare and the async exception will propagate and eventually close ChainDB so it won't hurt losing a couple of blocks if that's the case.